pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2018 Google Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type sockStatCollector struct {
+	inuse  *prometheus.Desc
+	tw     *prometheus.Desc
+	memory *prometheus.Desc
+}
+
+var pageSize = os.Getpagesize()
+
+func init() {
+	registerCollector("socket", NewSockStatCollector)
+}
+
+// NewSockStatCollector returns a new Collector exposing socket stats.
+func NewSockStatCollector() (Collector, error) {
+	return &sockStatCollector{
+		inuse: prometheus.NewDesc(
+			"num_inuse_sockets",
+			"Number of inuse sockets",
+			[]string{"protocol"}, nil,
+		),
+		tw: prometheus.NewDesc(
+			"num_tw_sockets",
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		memory: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+func initSocketStats() map[string]uint64 {
+	stats := map[string]uint64{
```

Comment:
why use a map with strings instead of just a struct?
```
type stats struct {
  tcpInUse uint64
  ...
}
```

Reply:
Because map is easier for automatic data assignment and integration.